### ISSUE TITLE
Setting CacheSignatureProviders = true cause exceptions

### ISSUE THREAD

**Issue by ycrumeyrolle** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#971]
Enabling the caching of the signature providers does not works in a multi-thread context. 

I have reproduced with this code:
```
            var keyData = new byte[32];
            RandomNumberGenerator.Fill(keyData);
            var handler = new JwtSecurityTokenHandler();
            var key = new SymmetricSecurityKey(keyData);
            key.CryptoProviderFactory.CacheSignatureProviders = true;
            Microsoft.IdentityModel.Logging.IdentityModelEventSource.ShowPII = true;
            var signingKey = new SigningCredentials(key, "HS256");
            Parallel.For(0, 10, _ =>
            {
                var token = handler.CreateJwtSecurityToken(issuer: "me", signingCredentials: signingKey);
                var jwt = handler.WriteToken(token);
            });
```

The exception is the following:
```
Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: Unknown error (0xc1000008)
   at Internal.Cryptography.HashProviderCng.TryFinalizeHashAndReset(Span`1 destination, Int32& bytesWritten)
   at Internal.Cryptography.HashProviderCng.FinalizeHashAndReset()
   at System.Security.Cryptography.HMACSHA256.HashFinal()
   at System.Security.Cryptography.HashAlgorithm.CaptureHashCodeAndReinitialize()
   at System.Security.Cryptography.HashAlgorithm.ComputeHash(Byte[] buffer)
   at Microsoft.IdentityModel.Tokens.SymmetricSignatureProvider.Sign(Byte[] input) in C:\agent2\_work\15\s\src\Microsoft.IdentityModel.Tokens\SymmetricSignatureProvider.cs:line 196
   at Microsoft.IdentityModel.JsonWebTokens.JwtTokenUtilities.CreateEncodedSignature(String input, SigningCredentials signingCredentials) in C:\agent2\_work\15\s\src\Microsoft.IdentityModel.JsonWebTokens\JwtTokenUtilities.cs:line 71
   at System.IdentityModel.Tokens.Jwt.JwtSecurityTokenHandler.WriteToken(SecurityToken token) in C:\agent2\_work\15\s\src\System.IdentityModel.Tokens.Jwt\JwtSecurityTokenHandler.cs:line 832
   at ConsoleApp2.Program.<>c__DisplayClass0_0.<Main>b__0(Int32 _) in c:\dev\Source\Repos\ConsoleApp2\ConsoleApp2\Program.cs:line 23
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
```

**Comment by brentschmaltz** [comment]
@ycrumeyrolle thanks.

**Comment by ycrumeyrolle** [comment]
Please find attached the project.
Tried with net461 without getting error, but fail with netcoreapp2.0 & 2.1. 

[ADAL-Issue971.zip](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/files/2280124/ADAL-Issue971.zip)

**Comment by mafurman** [comment]
@ycrumeyrolle I wasn't able to replicate this on either netcoreapp2.0 or netcoreapp2.1.

Do you get this error all the time or only sometimes?

**Comment by ycrumeyrolle** [comment]
Yes, I have exception with about 80% of runs.

**Comment by ycrumeyrolle** [comment]
Do you need some other information? Like DotNet sdk version?

**Comment by brentschmaltz** [comment]
@ycrumeyrolle dotnet version might come into play.

**Comment by brentschmaltz** [comment]
@ycrumeyrolle I was getting a different exception related to accessing memory after dispose.

**Comment by bartonjs** [comment]
`0xc1000008` might be trying to report ERROR_NOT_ENOUGH_MEMORY

**Comment by mafurman** [comment]
I'm closing this issue as it seems to be an issue with the cryptography libraries and not our library.

**Comment by jhudsoncedaron** [comment]
Turns out this bug is in fact in IdentityModel.JWT and still exists today in Microsoft.IdentityModel.JsonWebTokens. Sometimes, JsonWebTokenHandler.CreateToken() will use a *different* version of SecurityTokenDescriptor.SigningCredentials.Key than was passed to it via the initializer chain.

The problem is it second-guesses the call and despite being passed a fully constructed version of AsymmetricAlgorithm every time, this library tries to cache it, going around the `using` blocks and keeping references to dead native objects.

I've got this thing debugged and a reproduction in hand; unfortunately it will not bisect to a small reproduction.

The error manifests itself differently now since in the intervening years the code has been updated to check for calling disposed objects properly rather than trying to use dead native handles. 0xc1000008 is passing a bad handle to the native routines on Windows.

To make matters worse, CacheSignatureProviders is now defaulted to true.
